File: cpp/UnitTest/MultipleOverhangingTest.cpp
#include "pch.h"
#include "CppUnitTest.h"
#include "../Validator/ConstraintsLoading.h"
#include <string>
#include <vector>

using namespace Microsoft::VisualStudio::CppUnitTestFramework;
using namespace validator;

namespace LoadingConstraints {
	TEST_CLASS(MultipleOverhangingTest) {
		double alpha = 0.75;

	public:
		TEST_METHOD(FeasiblePosition) {
			// Arrangement consisting of two Towers
			// First Tower: Two Items
			// Second Tower: High Item
			// On Top: an Overhanging Item
			unsigned int dim = 6;

			ItemType typeNormal(1, dim, dim, dim,        0, false, 0);
			ItemType typeHigh  (2, dim, dim, 2 * dim,    0, false, 0);
			ItemType typeTop   (3, dim, 2 * dim +1, dim, 0, false, 0);

			Item itemNormal1(0, 1, 0, typeNormal);
			Item itemNormal2(1, 1, 0, typeNormal);
			Item itemHigh   (2, 1, 0, typeHigh);
			Item itemTop    (3, 1, 0, typeTop);

			// Instance Creation
			Customer customer(1, 0, 0, 4, 0, 0, 0, 0, 0);
			std::vector<Customer> customers{ customer };
			std::vector<ItemType> itemtypes{ typeNormal, typeHigh, typeTop };
			Instance instance("", Vehicle(0, 0, 0, 0, 0, 0, 0, 0, 0), itemtypes, customers, 0, 0);
			instance.items.emplace_back(itemNormal1);
			instance.items.emplace_back(itemNormal2);
			instance.items.emplace_back(itemHigh);
			instance.items.emplace_back(itemTop);

			// Tour Creation
			std::vector<unsigned int> customer_ids{ 1 };
			std::vector<unsigned int> item_ids{ 0, 1, 2, 3 };
			Tour tour(1, customer_ids, item_ids);

			// Positions
			instance.items.at(itemNormal1.id).setPosition(Point(0, 0, 0));
			instance.items.at(itemNormal2.id).setPosition(Point(0, 0, dim));
			instance.items.at(itemHigh.id)   .setPosition(Point(0, dim, 0));
			instance.items.at(itemTop.id)    .setPosition(Point(0, 0, 2* dim));

			// Tests
			const unsigned int endPos = tour.item_ids.size();
			ConstraintsLoading::getRelevantItems(instance.items.at(itemNormal1.id), tour, endPos, instance);
			ConstraintsLoading::getRelevantItems(instance.items.at(itemNormal2.id), tour, endPos, instance);
			ConstraintsLoading::getRelevantItems(instance.items.at(itemHigh.id),    tour, endPos, instance);
			ConstraintsLoading::getRelevantItems(instance.items.at(itemTop.id),     tour, endPos, instance);

			Assert::AreEqual(true, ConstraintsLoading::checkMultipleOverhanging(instance.items.at(itemTop.id), alpha, instance, true));
		}

		TEST_METHOD(InfeasiblePosition) {
			// Arrangement consisting of three stacked items,
			// which increase by 1/alpa per level

			unsigned int dim = 3;
			double w = 2 * dim;

			ItemType typeBottom(1, dim, w,                 dim, 0, false, 0);
			ItemType typeWide  (2, dim, w/alpha,           dim, 0, false, 0);
			ItemType typeTop   (3, dim, w/(alpha * alpha), dim, 0, false, 0);

			Item itemBottom(0, 1, 0, typeBottom);
			Item itemWide  (1, 1, 0, typeWide);
			Item itemTop   (2, 1, 0, typeTop);

			// Instance Creation
			Customer customer(1, 0, 0, 3, 0, 0, 0, 0, 0);
			std::vector<Customer> customers{ customer };
			std::vector<ItemType> itemtypes{ typeBottom, typeWide, typeTop };
			Instance instance("", Vehicle(0, 0, 0, 0, 0, 0, 0, 0, 0), itemtypes, customers, 0, 0);
			instance.items.emplace_back(itemBottom);
			instance.items.emplace_back(itemWide);
			instance.items.emplace_back(itemTop);

			// Tour Creation
			std::vector<unsigned int> customer_ids{ 1 };
			std::vector<unsigned int> item_ids{ 0, 1, 2 };
			Tour tour(1, customer_ids, item_ids);

			// Positions
			instance.items.at(itemBottom.id).setPosition(Point(0, 0, 0));
			instance.items.at(itemWide.id)  .setPosition(Point(0, 0, dim));
			instance.items.at(itemTop.id)   .setPosition(Point(0, 0, 2 * dim));

			const unsigned int endPos = tour.item_ids.size();
			ConstraintsLoading::getRelevantItems(instance.items.at(itemBottom.id), tour, endPos, instance);
			ConstraintsLoading::getRelevantItems(instance.items.at(itemWide.id),   tour, endPos, instance);
			ConstraintsLoading::getRelevantItems(instance.items.at(itemTop.id),    tour, endPos, instance);
			
			// Tests
			Assert::AreEqual(true, ConstraintsLoading::checkMinimalSupport(instance.items.at(itemWide.id), alpha, instance, true));
			Assert::AreEqual(true, ConstraintsLoading::checkMinimalSupport(instance.items.at(itemTop.id),  alpha, instance, true));

			Assert::AreEqual(false, ConstraintsLoading::checkMultipleOverhanging(instance.items.at(itemTop.id), alpha, instance, true));
		}
	};
}

File: java/src/test/java/com/threedimensionalloadingcvrp/validator/loading/MultipleOverhangingTest.java
package com.threedimensionalloadingcvrp.validator.loading;

import com.threedimensionalloadingcvrp.validator.model.*;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.Arrays;

import static com.threedimensionalloadingcvrp.validator.constraints.Loading.*;

public class MultipleOverhangingTest {
    float alpha = 0.75f;

    @Test
    public void FeasiblePosition() {
        // Arrangement consisting of two Towers
        // First Tower: Two Items
        // Second Tower: High Item
        // On Top: an Overhanging Item
        int dim = 6;

        ItemType typeNormal = new ItemType(1, dim, dim, dim,        0, false, 0);
        ItemType typeHigh   = new ItemType(2, dim, dim, 2 * dim,    0, false, 0);
        ItemType typeTop    = new ItemType(3, dim, 2 * dim +1, dim, 0, false, 0);

        Item itemNormal1 = new Item(0, 1, Rotation.none, typeNormal);
        Item itemNormal2 = new Item(1, 1, Rotation.none, typeNormal);
        Item itemHigh    = new Item(2, 1, Rotation.none, typeHigh);
        Item itemTop     = new Item(3, 1, Rotation.none, typeTop);

        // Instance Creation
        Instance instance = new Instance("", null, null, null, 0, false, Arrays.asList(itemNormal1, itemNormal2, itemHigh, itemTop));

        // Tour Creation
        Tour tour = new Tour(1, null, Arrays.asList(0, 1, 2, 3));

        // Positions
        instance.getItems().get(itemNormal1.getId()).setPosition(new Point(0, 0, 0));
        instance.getItems().get(itemNormal2.getId()).setPosition(new Point(0, 0, dim));
        instance.getItems().get(itemHigh.getId())   .setPosition(new Point(0, dim, 0));
        instance.getItems().get(itemTop.getId())    .setPosition(new Point(0, 0, 2* dim));

        // Tests
        int endPos = tour.getItem_ids().size();
        getRelevantItems(tour, endPos, instance);
        getRelevantItems(tour, endPos, instance);
        getRelevantItems(tour, endPos, instance);
        getRelevantItems(tour, endPos, instance);

        Assertions.assertTrue(checkMultipleOverhanging(instance.getItems().get(itemTop.getId()), alpha, instance, true));
    }

    @Test
    public void InfeasiblePosition() {
        // Arrangement consisting of three stacked items,
        // which increase by 1/alpa per level

        int dim = 3;
        int w = 2 * dim;

        ItemType typeBottom = new ItemType(1, dim, w,                 dim, 0, false, 0);
        ItemType typeWide   = new ItemType(2, dim, (int) (w/alpha),           dim, 0, false, 0);
        ItemType typeTop    = new ItemType(3, dim, (int) (w/(alpha * alpha)), dim, 0, false, 0);

        Item itemBottom = new Item(0, 1, Rotation.none, typeBottom);
        Item itemWide   = new Item(1, 1, Rotation.none, typeWide);
        Item itemTop    = new Item(2, 1, Rotation.none, typeTop);

        // Instance Creation
        Instance instance = new Instance("", null, null, null, 0, false, Arrays.asList(itemBottom, itemWide, itemTop));

        // Tour Creation
        Tour tour = new Tour(1, null, Arrays.asList(0, 1, 2));

        // Positions
        instance.getItems().get(itemBottom.getId()).setPosition(new Point(0, 0, 0));
        instance.getItems().get(itemWide.getId())  .setPosition(new Point(0, 0, dim));
        instance.getItems().get(itemTop.getId())   .setPosition(new Point(0, 0, 2 * dim));

		int endPos = tour.getItem_ids().size();
        getRelevantItems(tour, endPos, instance);
        getRelevantItems(tour, endPos, instance);
        getRelevantItems(tour, endPos, instance);

        // Tests
        Assertions.assertTrue(checkMinimalSupport(instance.getItems().get(itemWide.getId()), alpha, instance, true));
        Assertions.assertTrue(checkMinimalSupport(instance.getItems().get(itemTop.getId()),  alpha, instance, true));

        Assertions.assertFalse(checkMultipleOverhanging(instance.getItems().get(itemTop.getId()), alpha, instance, true));
    }
}
